package/MDAnalysis/analysis/parallel.py

```diff
@@ -0,0 +1,375 @@
+
+    .. _`scikit-learn`: https://scikit-learn.org/
+
+    Raises
+    ------
+    AttributeError
+        If an assigned attribute has the same name as a default attribute.
+
+    ValueError
+        If a key is not of type ``str`` and therefore is not able to be
+        accessed by attribute.
+
+    Examples
+    --------
+    >>> from MDAnalysis.analysis.base import Results
+    >>> results = Results(a=1, b=2)
+    >>> results['b']
+    2
+    >>> results.b
+    2
+    >>> results.a = 3
+    >>> results['a']
+    3
+    >>> results.c = [1, 2, 3, 4]
+    >>> results['c']
+    [1, 2, 3, 4]
+
+
+    .. versionadded:: 2.0.0
+    """
+
+    def _validate_key(self, key):
+        if key in dir(self):
+            raise AttributeError(f"'{key}' is a protected dictionary " "attribute")
+        elif isinstance(key, str) and not key.isidentifier():
+            raise ValueError(f"'{key}' is not a valid attribute")
+
+    def __init__(self, *args, **kwargs):
+        kwargs = dict(*args, **kwargs)
+        if "data" in kwargs.keys():
+            raise AttributeError(f"'data' is a protected dictionary attribute")
+        self.__dict__["data"] = {}
+        self.update(kwargs)
+
+    def __setitem__(self, key, item):
+        self._validate_key(key)
+        super().__setitem__(key, item)
+
+    def __setattr__(self, attr, val):
+        if attr == "data":
+            super().__setattr__(attr, val)
+        else:
+            self.__setitem__(attr, val)
+
+    def __getattr__(self, attr):
+        try:
+            return self[attr]
+        except KeyError as err:
+            raise AttributeError("'Results' object has no " f"attribute '{attr}'") from err
+
+    def __delattr__(self, attr):
+        try:
+            del self[attr]
+        except KeyError as err:
+            raise AttributeError("'Results' object has no " f"attribute '{attr}'") from err
+
+    def __getstate__(self):
+        return self.data
+
+    def __setstate__(self, state):
+        self.data = state
+
+
+class BackendBase:
+    """Base class for backend implementation. Initializes an instance and performs checks for its validity, such as n_workers and possibly other ones.
+
+    Parameters
+    ----------
+    n_workers : int
+        positive integer with number of workers (usually, processes) to split the work between
+
+    Examples
+    --------
+    >>> # implement a thread-based backend
+    >>> from MDAnalysis.analysis.parallel import BackendBase
+    >>> class ThreadsBackend(BackendBase):
+            def apply(self, func, computations):
+                from multiprocessing.dummy import Pool
+
+                with Pool(processes=self.n_workers) as pool:
+                    results = pool.map(func, computations)
+                return results
+    >>> from MDAnalysis.analysis.rms import RMSD
+    >>> R = RMSD(...) # setup the run
+    >>> n_workers = 2
+    >>> backend = ThreadsBackend(n_workers=n_workers)
+    >>> R.run(backend=backend)
```

Comment:
If the code actually runs, I'd keep the `>>>` (which you indeed need to have at each line). So we can make sure that problems are picked up by `doctest`. Otherwise what @yuxuanzhuang suggests is indeed better.